File: SpellStack/Frontend/SpellStack/src/utils/trialRules.ts
import type { TrialStarRequirement } from "../api/decks"

export const TRIAL_MINIMUM_DECK_WORD_COUNT = 10
export const TRIAL_ONE_STAR_THRESHOLD_PERCENT = 70
export const TRIAL_TWO_STAR_THRESHOLD_PERCENT = 80
export const TRIAL_MAXIMUM_STARS = 3

export function isDeckTrialEligible(wordCount: number) {
    return wordCount >= TRIAL_MINIMUM_DECK_WORD_COUNT
}

export function getTrialRequirements(totalQuestions: number): TrialStarRequirement[] {
    if (totalQuestions <= 0) return []

    return [
        {
            stars: 1,
            thresholdPercent: TRIAL_ONE_STAR_THRESHOLD_PERCENT,
            requiredCorrect: Math.ceil(totalQuestions * TRIAL_ONE_STAR_THRESHOLD_PERCENT / 100)
        },
        {
            stars: 2,
            thresholdPercent: TRIAL_TWO_STAR_THRESHOLD_PERCENT,
            requiredCorrect: Math.ceil(totalQuestions * TRIAL_TWO_STAR_THRESHOLD_PERCENT / 100)
        },
        {
            stars: TRIAL_MAXIMUM_STARS,
            thresholdPercent: 100,
            requiredCorrect: totalQuestions
        }
    ]
}

File: SpellStack/Backend/Services/TrialRules.cs
namespace SpellStack.Api.Services {
    public static class TrialRules {
        public const int MinimumDeckWordCount = 10;
        public const int OneStarThresholdPercent = 70;
        public const int TwoStarThresholdPercent = 80;
        public const int MaximumStars = 3;

        public static IReadOnlyList<TrialStarRequirement> GetRequirements(int totalQuestions) {
            if (totalQuestions <= 0) throw new ArgumentOutOfRangeException(nameof(totalQuestions));

            return [
                new TrialStarRequirement(1, OneStarThresholdPercent, RequiredCorrect(totalQuestions, OneStarThresholdPercent)),
                new TrialStarRequirement(2, TwoStarThresholdPercent, RequiredCorrect(totalQuestions, TwoStarThresholdPercent)),
                new TrialStarRequirement(MaximumStars, 100, totalQuestions)
            ];
        }

        public static int CalculateEarnedStars(int correctAnswers, int totalQuestions) {
            var requirements = GetRequirements(totalQuestions);
            return requirements
                .Where(requirement => correctAnswers >= requirement.RequiredCorrect)
                .Select(requirement => requirement.Stars)
                .DefaultIfEmpty(0)
                .Max();
        }

        private static int RequiredCorrect(int totalQuestions, int thresholdPercent) {
            return (int)Math.Ceiling(totalQuestions * thresholdPercent / 100d);
        }
    }

    public sealed record TrialStarRequirement(int Stars, int ThresholdPercent, int RequiredCorrect);
}
